sdk/data/azcosmos/cosmos_global_endpoint_manager.go

```diff
@@ -81,12 +81,12 @@ func (gem *globalEndpointManager) RefreshStaleEndpoints() {
 }
 
 func (gem *globalEndpointManager) ShouldRefresh() bool {
+	gem.gemMutex.RLock()
+	defer gem.gemMutex.RUnlock()
 	return time.Since(gem.lastUpdateTime) > gem.refreshTimeInterval
 }
 
 func (gem *globalEndpointManager) Update(ctx context.Context) error {
-	gem.gemMutex.Lock()
-	defer gem.gemMutex.Unlock()
```

Comment:
They are probably using this to protect other calls. Maybe audit GetAccountProperties and gem.locationCache.update for similar issues?

Another options would be to introduce an unlocked ShouldRefresh:

```go
func (gem *globalEndpointManager) ShouldRefresh() bool {
	gem.gemMutex.RLock()
	defer gem.gemMutex.RUnlock()
	return gem.shouldRefresh()
}

// shouldRefresh determines whether to refresh the endpoints. not threadsafe.
func (gem *globalEndpointManager) shouldRefresh() bool {
	return time.Since(gem.lastUpdateTime) > gem.refreshTimeInterval
}


func (gem *globalEndpointManager) Update(ctx context.Context) error {
        gem.gemMutex.Lock()
	defer gem.gemMutex.Unlock()
	if !gem.shouldRefresh() {
	// ...
	}
```

Reply:
hmm you are right, just saw that `lastUpdateTime` was updated inside `gem.locationCache.update` method. didn't see it is updated by `GetAccountProperties`, but I agree with what you said, your suggestion is safer if I keep the `Lock` at the beginning.